### PR TITLE
Dialog component

### PR DESCRIPTION
**Pull request by trickreich**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | yes
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | fixes --
| Related issues/PRs | --
| License | MIT
| Documentation PR | --

#### What's in this PR?

Dialog component.

#### Why?

We need dialog components for confirmation dialogs etc.

#### Example Usage

~~~js
initialState = {open: false};

const onConfirm = () => {
    /* do confirm things */
    setState({open: false});
};

const onCancel = () => {
    /* do cancel things */
    setState({open: false});
};

<div>
    <button onClick={() => setState({open: true})}>Open dialog</button>
    <Dialog
        title="Question?"
        onCancel={onCancel}
        onConfirm={onConfirm}
        cancelText="No"
        confirmText="Yes"
        open={state.open}>
        You have a question?
    </Dialog>
</div>
~~~

### CODE REVIEW

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Dialog/Actions.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
+// @flow
+import React from 'react';
+import type {Action} from './types';
+
+type Props = {
+    actions: Array<Action>,
+};
+
+export default class Actions extends React.PureComponent<Props> {
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Dialog/Actions.js]: Is it possible that this component is not used at all?

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Dialog/types.js [hunk 2]
```diff
@@ -0,0 +1,5 @@
+// @flow
+export type Action = {
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Dialog/types.js]: If the component above is not used we probably also don't need this type, right?

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Overlay/overlay.scss [hunk 3]
```diff
@@ -33,6 +33,7 @@ $transitionDuration: 300ms;
 }
 
 .overlay {
+    border-radius: 0 0 $borderRadius $borderRadius;
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Overlay/overlay.scss]: Was that a change brought to us by Tim?

File: src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/Datagrid.js [hunk 4]
```diff
@@ -79,13 +91,7 @@ export default class Datagrid extends React.PureComponent<Props> {
...
                     }
                 </div>
-                {!!page && !!pageCount &&
-                    <Pagination
-                        current={page}
-                        total={pageCount}
-                        onChange={this.handleChangePage}
-                    />
-                }
+                {pagination}
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/containers/Datagrid/Datagrid.js]: Why was that change necessary?

File: src/Sulu/Bundle/AdminBundle/Resources/js/containers/Form/registries/FieldRegistry.js [hunk 5]
```diff
@@ -21,6 +21,10 @@ class FieldRegistry {
     }
 
     get(name: string) {
+        if (!(name in this.fields)) {
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/containers/Form/registries/FieldRegistry.js]: I remember we've been talking about this. But please also add a test for that :-)

Comment (trickreich) [reply src/Sulu/Bundle/AdminBundle/Resources/js/containers/Form/registries/FieldRegistry.js]: Test is added

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Dialog/Dialog.js [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+                this.visible = this.props.open;
+            }
+        }));
+    }
+
+    @action handleTransitionEnd = () => {
+        afterElementsRendered(action(() => {
+            this.openHasChanged = false;
+        }));
+    };
```
[the remark this answers is not in the document]
Comment (trickreich) [reply src/Sulu/Bundle/AdminBundle/Resources/js/components/Dialog/Dialog.js]: We need to implement the correct animation

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Dialog/dialog.scss [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    transition: top $transitionDuration, opacity $transitionDuration;
+    z-index: 1;
+    pointer-events: none;
+
+    & > * {
+        pointer-events: auto;
+    }
+}
+
+.is-open {
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Dialog/dialog.scss]: Sry, seems like I have overseen that one before... should be just `open`.

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Dialog/dialog.scss [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+$titleColor: $black;
+$contentColor: $dustyGray;
+
+$confirmBackground: $shakespeare;
+$confirmColor: $white;
+
+$transitionDuration: 300ms;
+
+.container {
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Dialog/dialog.scss]: Sry, now that we've been talking about that, I've seen that we have the same situation here 🙈 Would go for `dialog-container`.

File: src/Sulu/Bundle/AdminBundle/Resources/js/components/Button/Button.js [hunk 9]
```diff
@@ -0,0 +1,31 @@
+// @flow
+import React from 'react';
+import classNames from 'classnames';
+import buttonStyles from './button.scss';
+
+type Props = {
+    children: string,
+    type: 'primary' | 'secondary',
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/components/Button/Button.js]: In other cases we called this variable `skin`.
